Server/mods/deathmatch/utils/CZipMaker.cpp

```diff
@@ -133,35 +164,112 @@ bool CZipMaker::InsertDirectoryTree(const SString& strInSrc, const SString& strI
+{
+    char* entry[] = { const_cast<char*>(strPath.c_str()) };
+    return zip_entries_delete(m_uzFile, entry, 1) == 1;
+}
+
+bool CZipMaker::RemoveFile(const CZipMaker::CZipEntry& entry) noexcept
+{
+    return RemoveFile(entry.name);
+}
+
+bool CZipMaker::Replace(const SString& strSrc, const SString& strDest) noexcept
+{
+    if (!RemoveFile(strSrc))
+        return false;
+    if (!InsertFile(strDest))
+        return false;
+
+    return true;
+}
+bool CZipMaker::Exists(const SString& strPath) const noexcept
+{
+    bool bStatus = zip_entry_open(m_uzFile, strPath) == 0;
+    return bStatus && zip_entry_close(m_uzFile) == 0;
+}
+bool CZipMaker::ExtractFile(const SString& strFilePath, const SString& strDirPath) noexcept
 {
-    // save file attributes
-    zip_fileinfo zfi;
+    if (!m_uzFile)
+        return false;
 
-    zfi.internal_fa = 0;
-    zfi.external_fa = 0;
+    if (zip_entry_open(m_uzFile, strFilePath))
+        return false;
 
-    // save file time
-    time_t secondsNow = time(NULL);
-    tm*    tmp = gmtime(&secondsNow);
+    auto destPath = PathJoin(strDirPath, strFilePath);
+    MakeSureDirExists(destPath);
+    if (zip_entry_fread(m_uzFile, destPath))
+    {
+        zip_entry_close(m_uzFile);
+        return false;
+    }
 
-    zfi.dosDate = 0;
-    zfi.tmz_date.tm_year = tmp->tm_year + 1900;
-    zfi.tmz_date.tm_mon = tmp->tm_mon;
-    zfi.tmz_date.tm_mday = tmp->tm_mday;
-    zfi.tmz_date.tm_hour = tmp->tm_hour;
-    zfi.tmz_date.tm_min = tmp->tm_min;
-    zfi.tmz_date.tm_sec = tmp->tm_sec;
+    return zip_entry_close(m_uzFile) == 0;
+}
+bool CZipMaker::Extract(const SString& strDirPath) noexcept
+{
+    if (!m_uzFile)
+        return false;
 
-    int iResult = zipOpenNewFileInZip(m_uzFile, strDest, &zfi, NULL, 0, NULL, 0, NULL, Z_DEFLATED, Z_DEFAULT_COMPRESSION);
+    return zip_extract(m_strZipPath, strDirPath, NULL, NULL) == 0;
+}
 
-    if (iResult == ZIP_OK)
+std::vector<CZipMaker::CZipEntry> CZipMaker::ListEntries() const noexcept
+{
+    std::vector<CZipEntry> entries;
+
+    for (auto i = 0; i < zip_entries_total(m_uzFile); i++)
     {
-        if (!buffer.empty())
-            iResult = zipWriteInFileInZip(m_uzFile, &buffer[0], buffer.size());
+        entries.push_back(GetFileByIndex(i));
     }
 
-    zipCloseFileInZip(m_uzFile);
+    return entries;
+}
+
+CZipMaker::CZipEntry CZipMaker::operator[](unsigned long long offset) const noexcept
+{
+    return GetFileByIndex(offset);
+}
+CZipMaker::operator bool() const noexcept
+{
+    return IsValid();
+}
+
+CZipMaker::CZipEntry CZipMaker::GetFileByIndex(unsigned long long offset) const noexcept
+{
+    CZipEntry entry;
+    if (!m_uzFile)
+        return entry;
+
+    const auto size = zip_entries_total(m_uzFile);
+    if (offset > size)
+        return entry;
+
+    if (offset < 0)
+        offset += size;
+
+    if (zip_entry_openbyindex(m_uzFile, offset))
+        return entry;
+
+    entry.name = zip_entry_name(m_uzFile);
```

Comment:
It is unsafe to store a pointer to a string this way. When zip file get closed all these strings become invalid. And I can see that it is already a problem in here since m_vecEntries won't be cleared when ::Close called.